### PR TITLE
Add Create Prefab

### PR DESCRIPTION
**Pull request by crystaldev3d**
Hi! I think this would also be a useful menu option for the editor.

### CODE REVIEW

File: Source/Editor/Foundation/SceneViewTab.cpp [hunk 1]
```diff
@@ -511,6 +516,37 @@ void SceneViewTab::FocusSelection(SceneSelection& selection)
...
+
+                Vector3 oldPrefabPosition = activeNode->GetWorldPosition();
+                activeNode->SetWorldPosition(Vector3::ZERO);
+                FileSystem* fs = GetSubsystem<FileSystem>();
+
+                if (fs)
+                {
+                    eastl::string path = fs->FindResourcePrefixPath();
+                    File file = File(context_, path + "Data/Prefabs/" + prefabFileName, FILE_WRITE);
+                    activeNode->SaveXML(file);
```
Comment (rokups) [inline Source/Editor/Foundation/SceneViewTab.cpp:542]: @eugeneko should we not use new serialization here?

File: Source/Editor/Foundation/SceneViewTab.cpp [hunk 2]
```diff
@@ -511,6 +516,37 @@ void SceneViewTab::FocusSelection(SceneSelection& selection)
...
+    if (Node* activeNode = selection.GetActiveNode())
+    {
+        if (SceneViewPage* page = GetPage(activeNode->GetScene()))
+        {
+            if (activeNode)
+            {
+                eastl::string prefabFileName = activeNode->GetName();
+
+                if (prefabFileName.length() < 1)
+                    prefabFileName = prefabFileName.sprintf("%s %d", "PrefabNodeID_", activeNode->GetID());
```
Comment (rokups) [inline Source/Editor/Foundation/SceneViewTab.cpp:530]: Use `Format()` function for string formatting. `Format("PrefabNodeID_{}", activeNode->GetID())`.

File: Source/Editor/Foundation/SceneViewTab.cpp [hunk 3]
```diff
@@ -511,6 +516,37 @@ void SceneViewTab::FocusSelection(SceneSelection& selection)
...
 }
+void SceneViewTab::CreatePrefabFile(SceneSelection& selection)
+{
+    if (Node* activeNode = selection.GetActiveNode())
+    {
+        if (SceneViewPage* page = GetPage(activeNode->GetScene()))
+        {
+            if (activeNode)
+            {
+                eastl::string prefabFileName = activeNode->GetName();
```
Comment (rokups) [inline Source/Editor/Foundation/SceneViewTab.cpp:527]: Unnecessary string copy.

**Comment by eugeneko** [comment]
I need to think about prefabs in Editor for a while.
Thanks for PR, I'll consider this code.

**Comment by eugeneko** [comment]
All in all, I like this prefab, but I have one issue with it: this kind of code should be implemented as SceneViewAddon instead of expanding main file. I can do it myself when I have time, or you can try doing it yourself.

`DragAndDropPrefabsToSceneView` is moved to addon trivially.

`CreatePrefabFile` can be moved to addon by adding one more virtual function to SceneViewAddon, or one more signal to SceneViewTab

PS: I would like `CreatePrefabFile` to use create prompt, similar to "Create Scene", but it may need some refactoring. I will look into it later. I am busy with UI integration now.

**Comment by crystaldev3d** [comment]
ok I'll try to implement this as addon.

**Comment by eugeneko** [comment]
It's my turn now, let me take a look and then I will merge...

**Comment by crystaldev3d** [comment]
Okay, but I still haven't figured out how to save the changes that make the drag and drop (

**Comment by eugeneko** [comment]
I updated the dialog for prefab creation, it should be more consistent now with other dialogs.

**Comment by eugeneko** [comment]
@crystaldev3d 
> but I still haven't figured out how to save the changes that make the drag and drop (

Can you elaborate please?

**Comment by crystaldev3d** [comment]
For example, the material assignment by drag&drop is not saved when you reload the editor. 
Even if you clicked before save the project.

**Comment by eugeneko** [comment]
@crystaldev3d can you check/test this PR one last time? I think I am done with it.

**Comment by crystaldev3d** [comment]
@eugeneko great work, It's works almost perfectly, but I can't update the structure of a previously created prefab(( 
because creating a prefab is often an iterative process, why should it be limited to one attempt?)
Clipboard01

**Comment by crystaldev3d** [comment]
You could also save the path where the prefab was saved (Node->Vars?) and show the saving path in the file field when the prefab is updated, so the user doesn't get confused and save the same prefab in different folders. I had this situation because the wrong folder was selected in the resource tab.

**Comment by eugeneko** [comment]
@crystaldev3d thanks for the feedback!

I understand your use case. The reason it works this way is that file operations cannot be undone, so it's pretty dangerous to allow user "just overwrite" something.

In perfect world, your use case should be covered by inline prefab editing: when you edit _instantiated_ prefab as part of the scene, you should be able to just save it from inspector.

Having said that, why don't you edit your prefab as is? I mean, just open your prefab and just edit it, instead of re-saving it from scene? If this is unsuitable for you, let me know and I will add an option to enable file overwrite.

**Comment by crystaldev3d** [comment]
@eugeneko Yes it would be great to be able to change the prefab directly from the same opened scene. 
Instead of opening it separately. It's just more convenient.
Just need an additional option when saving - "allow overwriting prefab"

**Pull request merged**